### PR TITLE
[TPU] Optimize kv cache update kernel

### PR DESCRIPTION
**Pull request by tengyifei**
## Purpose

This PR improves the throughput of the kv cache update kernel from 28.15 GB/s to 91.65 GB/s on v6e.

This PR adds to optimizations on top of https://github.com/vllm-project/vllm/pull/19928. It picks the optimal number of slices to copy per kernel program instance based on results from microbenchmarks.

Reference data demonstrating improvements: https://github.com/tengyifei/playground/blob/master/pallas/better-index-copy.ipynb

An earlier commit in this PR also pipelined the DMAs into and out of VMEM, resulting in more in-flight DMAs. That improved throughput to 100 GB/s in microbenchmarks but actually _decreased_ the performance in `python3 ./benchmarks/benchmark_serving.py`. I'm not sure why this happens. In any case, I omitted that optimization in the latest commit because what matters is the end-to-end benchmark.

## Test plan

Kernel test: pytest -s -v tests/v1/tpu/test_kv_cache_update_kernel.py
Accuracy test: pytest -s -v tests/entrypoints/llm/test_accuracy.py::test_lm_eval_accuracy_v1_engine

## Test result

PASSED

## Llama 3.1 70B perf

```
vllm serve meta-llama/Llama-3.1-70B-Instruct --disable-log-requests --gpu-memory-utilization 0.98 --max-num-batched-tokens 2048 --max-num-seqs 128 --max-model-len 2048 --no-enable-prefix-caching --tensor_parallel_size=8

python3 ./benchmarks/benchmark_serving.py --model meta-llama/Llama-3.1-70B-Instruct --dataset-name sonnet --dataset-path benchmarks/sonnet_4x.txt --sonnet-input-len 1800 --sonnet-output-len 128 --ignore_eos
```

- Before: 6.73 reqs/s
- This PR: 6.77 reqs/s

### CODE REVIEW

**Comment by tengyifei** [comment]
cc @yaochengji

File: vllm/v1/worker/tpu_model_runner.py [hunk 1]
```diff
@@ -590,6 +595,10 @@ def _get_slot_mapping_metadata(self, num_reqs,
             [kv_cache_start_indices, new_kv_start_indices, slice_lens], axis=1)
         return slot_mapping_metadata
 
+    def _get_page_size_bytes(self):
```
Comment (yaochengji) [inline vllm/v1/worker/tpu_model_runner.py]: I'd like to put this util function in pallas.py

File: vllm/v1/worker/tpu_model_runner.py [hunk 2]
```diff
@@ -1807,19 +1817,41 @@ def _get_padded_token_len(paddings: list[int], x: int) -> int:
...
+    will increase HBM bandwidth utilization via more in-flight DMAs.
+
+    However, it will also use more VMEM, and experimentally, we observed
+    performance regression at 128 slices on v6e, likely due to running
+    out of scalar registers. Thus this function will limit the number of
+    slices to 64.
+    """
+    # Conservative VMEM usage limit: 32 MiB
+    vmem_limit = 32 * 1024 * 1024
+    num_slices_per_block = vmem_limit // page_size_bytes
```
Comment (yaochengji) [inline vllm/v1/worker/tpu_model_runner.py:1849]: If it's double buffering, I think we need to make it `vmem_limit // page_size_bytes // 2`

File: vllm/v1/worker/tpu_model_runner.py [hunk 3]
```diff
@@ -1807,19 +1817,41 @@ def _get_padded_token_len(paddings: list[int], x: int) -> int:
...
+    assert num_slices_per_block > 0, "Number of slices should be positive"
+    num_slices_per_block = next_power_of_2(num_slices_per_block)
```
Comment (yaochengji) [inline vllm/v1/worker/tpu_model_runner.py]: Why is it `next_power_of_2` instead of `previous_power_of_2`?

**Review by yaochengji — COMMENTED** [review]
Thanks for your contribution, Yifei! Left a few comments.

File: vllm/v1/worker/tpu_model_runner.py (hunk 3 above)
Comment (tengyifei) [reply vllm/v1/worker/tpu_model_runner.py]: No particular reason. `prev_power_of_2` seems safer.

File: vllm/v1/worker/tpu_model_runner.py (hunk 2 above)
Comment (tengyifei) [reply vllm/v1/worker/tpu_model_runner.py:1849]: Based on my understanding of the `vllm.attention.ops.pallas_kv_cache_update` kernel, Pallas won't emit an overlapped pipeline here, because the input lives in `ANY` instead of `VMEM`, c.f. https://docs.jax.dev/en/latest/pallas/tpu/pipelining.html#tpu-memory-spaces. So the kernel would just copy inputs from HBM to VMEM, wait for that, then copy from VMEM to output HBM, wait for that, and repeat.

File: vllm/v1/worker/tpu_model_runner.py (hunk 1 above)
Comment (tengyifei) [reply vllm/v1/worker/tpu_model_runner.py]: Done

File: vllm/v1/worker/tpu_model_runner.py (hunk 2 above)
Comment (yaochengji) [reply vllm/v1/worker/tpu_model_runner.py:1849]: Yeah, what I mean is that when you're trying to change it to use double buffering, we need to have an additional `// 2`.

**Review by yaochengji — APPROVED** [review]
LGTM, thanks! 

Could you sign the DCO?

**Comment by mgoin** [comment]
No need to sign the DCO as we can override, but please fix the pre-commit

**Comment by tengyifei** [comment]
> Could you sign the DCO?

Done

> please fix the pre-commit

Done
